pipeline_api/src/main/java/com/splicemachine/pipeline/writehandler/UpdateUtils.java

```diff
@@ -0,0 +1,84 @@
+package com.splicemachine.pipeline.writehandler;
+
+import com.carrotsearch.hppc.BitSet;
+import com.carrotsearch.hppc.BitSetIterator;
+import com.splicemachine.encoding.MultiFieldDecoder;
+import com.splicemachine.kvpair.KVPair;
+import com.splicemachine.storage.EntryDecoder;
+import com.splicemachine.storage.index.BitIndex;
+import com.splicemachine.storage.index.BitIndexing;
+import com.splicemachine.utils.ByteSlice;
+
+public class UpdateUtils {
+
+    public static KVPair deleteFromUpdate(KVPair mutation) {
+        EntryDecoder rowDecoder = new EntryDecoder();
+        rowDecoder.set(mutation.getValue());
+        BitIndex index = rowDecoder.getCurrentIndex();
+        MultiFieldDecoder fieldDecoder = rowDecoder.getEntryDecoder();
+        int cardinality = index.cardinality();
+        for(int i=index.nextSetBit(0), c=0; c<cardinality/2; i=index.nextSetBit(i+1), c++){
+            rowDecoder.seekForward(fieldDecoder, i);
+        }
+        byte[] data = fieldDecoder.slice(fieldDecoder.length() - fieldDecoder.offset());
+        BitIndex resultIndex = BitIndexing.getBestIndex(
+                halveSet(index.getFields()),
+                halveSet(index.getScalarFields()),
+                halveSet(index.getFloatFields()),
+                halveSet(index.getDoubleFields()));
+        byte[] bitData = resultIndex.encode();
+        byte[] result = new byte[bitData.length+data.length+1];
+        System.arraycopy(bitData, 0, result, 0, bitData.length);
+        result[bitData.length] = 0;
+        System.arraycopy(data,0,result,bitData.length+1,data.length);
+
+        ByteSlice rowSlice = mutation.rowKeySlice();
+        return new KVPair(
+                rowSlice.array(),rowSlice.offset(),rowSlice.length(),
+                result,0,result.length,KVPair.Type.DELETE);
+    }
+
+    public static KVPair getBaseUpdateMutation(KVPair mutation) {
+        EntryDecoder rowDecoder = new EntryDecoder();
+        rowDecoder.set(mutation.getValue());
+        BitIndex index = rowDecoder.getCurrentIndex();
+        MultiFieldDecoder fieldDecoder = rowDecoder.getEntryDecoder();
+        int offset = fieldDecoder.offset();
+        for(int i=index.nextSetBit(0), c=0; c<index.cardinality()/2; i=index.nextSetBit(i+1), c++){
+            rowDecoder.seekForward(fieldDecoder, i);
+        }
+        int end = fieldDecoder.offset();
+        fieldDecoder.seek(offset);
+        byte[] data = fieldDecoder.slice(end - offset - 1); // -1 to remove the NULL separator
```

Comment:
(copied from https://github.com/splicemachine/spliceengine/pull/5384)

```
int offset = fieldDecoder.offset();
int end = fieldDecoder.offset();
byte[] data = fieldDecoder.slice(end - offset - 1);
```
this looks like
`byte[] data = fieldDecoder.slice( - 1);`
????
am i missing something?

why is this so different from line 23?
`byte[] data = fieldDecoder.slice(fieldDecoder.length() - fieldDecoder.offset());`

Apart from this, this function looks like a copy&paste of the function deleteFromUpdate, except for KVPair.Type.UPDATE

Reply:
ah ok `rowDecoder.seekForward(fieldDecoder, i);` would change this. But why is that code so different from deleteFromUpdate?

Reply:
In order to decode the value we need to skip over one field at a time. In general fields are separated by a null value `0x00` but depending some types of fields (scalar, float & double) can contain null bytes, and the type information is encoded in the BitIndex structure.
So if we know we have 4 fields, we skip over 2 fields and get that offset. If we want the second half we take everything from that point till the end, if we want the first half remember that position, we `seek` the decoder back to the beginning and take everything from the beginning to the middle position.
I renamed some variables to hopefully make it more understandable.